### PR TITLE
Simplify node shutdown handling.

### PR DESCRIPTION
**Pull request by idegtiarenko**
This adds a dedicated method to retrieve shutdown metadata and check its type at once 
as well as removes unnecessary optional handling of metadata.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/NodesShutdownMetadata.java [hunk 1]
```diff
@@ -67,27 +67,6 @@ public static NamedDiff<Metadata.Custom> readDiffFrom(StreamInput in) throws IOE
...
-        // Right now we make no distinction between the type of shutdown, but maybe in the future we might?
-        return NodesShutdownMetadata.getShutdowns(state)
-            .map(NodesShutdownMetadata::getAllNodeMetadataMap)
-            .map(allNodes -> allNodes.get(nodeId))
-            .isPresent();
-    }
-
-    public static NodesShutdownMetadata getShutdownsOrEmpty(final ClusterState state) {
-        return getShutdowns(state).orElse(EMPTY);
-    }
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/metadata/NodesShutdownMetadata.java:89]: This is replaced with `org.elasticsearch.cluster.metadata.Metadata#nodeShutdowns` that never returns null

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SnapshotInProgressAllocationDecider.java [hunk 2]
```diff
@@ -59,7 +59,7 @@ private static Decision canMove(ShardRouting shardRouting, RoutingAllocation all
             return YES_NOT_SNAPSHOTTED;
         }
 
-        SnapshotsInProgress snapshotsInProgress = allocation.custom(SnapshotsInProgress.TYPE);
+        SnapshotsInProgress snapshotsInProgress = allocation.getClusterState().custom(SnapshotsInProgress.TYPE);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SnapshotInProgressAllocationDecider.java:62]: Why this change?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java [hunk 3]
```diff
@@ -39,7 +39,7 @@ public Decision canAllocate(final ShardRouting shardRouting, final RoutingAlloca
             return allocation.decision(Decision.YES, NAME, "not an API-level restore");
         }
 
-        final RestoreInProgress restoresInProgress = allocation.custom(RestoreInProgress.TYPE);
+        final RestoreInProgress restoresInProgress = allocation.getClusterState().custom(RestoreInProgress.TYPE);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java:42]: Why this change?

File: server/src/main/java/org/elasticsearch/cluster/metadata/NodesShutdownMetadata.java [hunk 4]
```diff
@@ -106,10 +85,45 @@ public void writeTo(StreamOutput out) throws IOException {
...
+     */
+    @Nullable
+    public SingleNodeShutdownMetadata get(String nodeId) {
+        return nodes.get(nodeId);
+    }
+
+    /**
+     * Returns true if the given node is marked as shutting down with any shutdown type.
+     */
+    public boolean contains(String nodeId) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/metadata/NodesShutdownMetadata.java:110]: NIT: Maybe retain the naming of `isShuttingDown` here? `contains` doesn't really describe the method well?

**Review by original-brownbear — APPROVED** [review]
LGTM but seems two unrelated changes (that don't really matter either way) snuck in here?
Also maybe fix the naming on `contains` :)
Either way fine to merge though, definitely an improvement over the existing code.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java (hunk 3 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java:42]: I removed it to make RoutingAllocation simpler and not to mirror a lot of cluster state calls

File: server/src/main/java/org/elasticsearch/cluster/metadata/NodesShutdownMetadata.java (hunk 4 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/metadata/NodesShutdownMetadata.java:110]: Typical call would already include shutdown in it, I would like to avoid repeating that: `state.metadata().nodeShutdowns().contains(nodeId)`.
I also wanted not to use `shutting down` here because of it is possible confusion with `RESTARTING` type that could be returned by this call.

**Comment by idegtiarenko** [comment]
@elasticsearchmachine please run elasticsearch-ci/bwc

**Comment by pxsalehi** [comment]
@idegtiarenko I think this broke the new serverless-sigterm module and `SigtermShutdownCleanupService` doesn't compile. See https://buildkite.com/elastic/elasticsearch-check-serverless-submodule/builds/5
